src/sgame/sg_bot_nav.cpp

```diff
@@ -368,6 +368,51 @@ Local Bot Navigation
 ========================
 */
 
+bool BotTraceForFloor( gentity_t *self, uint32_t dir )
+{
+	// this is using a daemon trace to a solid surface
+	// another feasible method might be to do a recast navmesh trace
+	trace_t trace;
+	glm::vec3 dirVec;
+	glm::vec3 viewangles = VEC2GLM( self->client->ps.viewangles );
+	viewangles.x = 0;
+	AngleVectors( viewangles, &dirVec, nullptr, nullptr );
+	// we only handle a subset of the possible values of the bitfield `dir`
+	if ( dir == MOVE_BACKWARD )
+	{
+		dirVec.x = -dirVec.x;
+		dirVec.y = -dirVec.y;
+	}
+	else if ( dir == MOVE_LEFT )
+	{
+		std::swap( dirVec.x, dirVec.y );
+		dirVec.x = -dirVec.x;
+	}
+	else if ( dir == MOVE_RIGHT )
+	{
+		std::swap( dirVec.x, dirVec.y );
+		dirVec.y = -dirVec.y;
+	}
+	else
+	{
+		return false;
+	}
+	glm::vec3 playerMins, playerMaxs;
+	class_t pClass = static_cast<class_t>( self->client->ps.stats[STAT_CLASS] );
+	BG_BoundingBox( pClass, &playerMins, &playerMaxs, nullptr, nullptr, nullptr );
+	float radius = playerMaxs.x - playerMins.x;
+	float height = playerMaxs.z - playerMins.z;
+	// use dirVec * ( 2.f * radius ) as direction offset
+	// in theory, sqrtf(2.f) instead of 2.f would be enough
+	// the slight increase is to safeguard against unexpected map geometry
+	// a bot not dancing is less problematic than a bot falling down
+	glm::vec3 start = VEC2GLM( self->s.origin ) + dirVec * ( 2.f * radius );
+	glm::vec3 end = start;
+	end.z -= height;
+	trap_Trace( &trace, &start[ 0 ], nullptr, nullptr, &end[ 0 ], self->num(), MASK_SOLID, 0 );
```

Comment:
Use the GLM variant: replace `&start[0], nullptr, nullptr, &end[0]` with `start, {}, {}, end`

Reply:
Do we have any guidelines on how and when and why to use the glm variants? I am lost here. And I bet I'm not the only one.

Reply:
The GLM variants just have the first 4 arguments as glm::vec3 instead of vec3_t. So you should definitely use them if all your vectors are glm::vec3.

But come to think of it, you might want to use `G_Trace2` instead of `trap_Trace`. This is because trap_Trace may ignore entities that overlap the starting point of the trace, which is probably not desired here. G_Trace2 does not have a GLM variant.

Reply:
No thx. Your glm migration is beyond me.